### DISCUSSION TITLE
How to organize single Dagster repo with multiple code locations and DBT projects

### DISCUSSION THREAD

**Discussion by stevenmurphy12** [dagster-io/dagster#18163] [Q&A] [answered]
![image](https://github.com/dagster-io/dagster/assets/6832968/17bdbdb2-9e6a-4ba7-9d7c-8d83e63faf61)

I have a single Git repository which contains multiple code locations. Each code location represents the assets for one of our clients. Each client requires assets derived from both Python code and DBT code. I expect the amount of DBT models to be quite low, couple of dozen at most.

My current setup is depicted in Option 1: separate DBT repo for each client. Until now I've been doing a proof-of-concept on a single client, and reproducing the set up to another client has felt quite heavyweight, and prone to misconfiguration. 

I'm considering option 2, by which I have a single DBT project that contains all the DBT models required for all clients, and each code location pulls in the DBT models it requires through appropriate use of the `@dbt_assets` decorator and `select/exclude`.

Is their precedent for either option? I'm leaning towards Option 2, whilst it breaks encapsulation it would be less set up for onboarding new clients. 

The reason each client has their own code location is to enable better separation of their assets in the UI (as code locations give a another level of grouping), it's possible that UI modifications may make this point moot, as I'd be satisfied with everything in a single code location if the UI would facilitate it - https://github.com/dagster-io/dagster/discussions/16657#discussioncomment-7622296

**Comment by benpankow** [comment, marked as answer]
Hi Steven,

Right now there are roughly two levels of grouping, as you point out: code locations, which define different execution environments & also act as collapsible sets of definitions in the UI, and asset groups which show up in the asset graph as collapsible.

The second option seems reasonable to me, especially if you limit the set of assets you're loading in each code location with selection.  We use this pattern internally where DBT models corresponding to different use-cases are colocated in the same DBT project. You could [use a custom DBT translator or metadata to assign asset groups](https://docs.dagster.io/integrations/dbt/reference#customizing-group-names) which could let you subdivide by client or function.

We've been exploring concepts for nested groups, but don't yet have a concrete proposal/roadmap on this end.

**Reply by stevenmurphy12** [reply]
Hi @benpankow ,

> We've been exploring concepts for nested groups, but don't yet have a concrete proposal/roadmap on this end.

Have you/your colleagues came to a conclusion on the nested groups idea? 

As of writing we've got around 30 customers set up on Dagster, following the pattern of a single code location per customer. This is resulting in deployments of updated definitions taking longer as Kubernetes has to spin up all those pods, plus the Dagster webserver/daemon need to go through an init process for each one. These pods are also taking up resources which seems wasteful. I'm anticipating around 90 customers in the medium term which will compound the issue.

I got into a good discussion [here](https://dagster.slack.com/archives/C014N0PK37E/p1721224065222519) in which the use of `create_repository_using_definitions_args` was recommended. This is an `@experimental` function. I moved a couple of code locations into it and functionally it seems OK, but it has the following side effects:
1. Have to re-enable sensors/schedules (as their default run status is `STOPPED`). I'm guessing it's something to do with the code location name: `customername@codelocname` as opposed to `customername`
2. I can't view the run history of jobs that I've moved into the singular code location
3. Seeing a lot of DB errors relating to Dagster trying to operate on closed connections (see the Slack thread). It appears to have no functional effect, but the many errors are worrying.

For now I'm planning to revert back to my previous approach (one code location per customer), and hopefully wait for the grouping functionality.

For the grouping functionality: could that result in the asset keys changing? i'd like to keep them as they are if possible for the pipelines already implemented.

Thank you

**Reply by garethbrickman** [reply]
Nested groups is not currently on the development roadmap. Here is the tracking issue for it: https://github.com/dagster-io/dagster/issues/14530

**Reply by stevenmurphy12** [reply]
Thank you for the info.

**Comment by stevenmurphy12** [comment]
Thanks Ben, I've switched over to using a single DBT project across two code locations for now. Also using a DBT custom translator to specify the group name (based on the directory hierarchy, so if I had `myproject_dbt/models/client_one/pipeline_one/mymodel.sql`, the group becomes `client_one_pipeline_one`)

```python

class CustomDagsterDbtTranslator(DagsterDbtTranslator):
    def get_group_name(self, dbt_resource_props: Mapping[str, Any]) -> Optional[str]:
        fqn = dbt_resource_props["fqn"]
        group_name = fqn[1] + "_" + fqn[2]

        return group_name
```

I also want the data for each client to be in separate BigQuery datasets, I've achieved this by having a `client_name` passed into the the DBT commands, that specifies the dataset at runtime.

```yaml
myproject_dbt:
  outputs:
    staging:
      dataset: staging_{{ var('client_name') }}_transform
      job_execution_timeout_seconds: 300
      job_retries: 1
      location: a_region
      priority: interactive
      project: gcp_project
      threads: 1
      type: bigquery
      method: oauth
    production:
      dataset: production_{{ var('client_name') }}_transform
      job_execution_timeout_seconds: 300
      job_retries: 1
      location: a_region
      priority: interactive
      project: gcp_project
      threads: 1
      type: bigquery
      method: oauth
  target: staging
```

Do both of these seem reasonable?

Tangentially related: each time the `DbtCliResource` is invoked it outputs the contents to a unique directory (by design though it can be overridden - https://dagster.slack.com/archives/C04CW71AGBW/p1700521803566509?thread_ts=1700506910.047309&cid=C04CW71AGBW). I'm guessing you've been running this for a while, have you had any issues with all the folders building up over time? I can imagine them building up. Though a counter-point would be that since (in my case), I'm running these jobs on Kubernetes pods they're short lived anyway (as I was writing this I was thinking of my experience in my local dev environment).

**Reply by benpankow** [reply]
Yes, both these approaches make sense to me!

> have you had any issues with all the folders building up over time?

Yes, generally it hasn't been an issue - we've `.gitignore`d them for version control purposes, and the actual production pods are short-lived enough that it hasn't been a problem.

**Reply by stevenmurphy12** [reply]
Cool thank you

**Comment by stevenmurphy12** [comment]
An addendum to the dataset name override:

I started out building my Dagster/DBT integration with the following example: https://docs.dagster.io/integrations/dbt/using-dbt-with-dagster/load-dbt-models#step-4-understand-the-python-code-in-your-dagster-project. In here you use `DAGSTER_DBT_PARSE_PROJECT_ON_LOAD` to determine whether to do a DBT parse to build a manifest. 

For local development it is on; for a deployment to Kubernetes I've turned it off, and pre-parsed the DBT project in my Dockerfile. A manifest is built per target environment. This is now failing because it is expecting the `client_name` parameter. 

I could do a `dbt parse` command for each client. Though this would mean a manifest per environment, per client, which I can see growing quite quickly and becoming a minor maintenance headache.

I was wondering whether it would be worth just letting Dagster do a build of the manifest each time by switching on `DAGSTER_DBT_PARSE_PROJECT_ON_LOAD`.

 Convenience over efficiency, any thoughts on it?
